docker/Makefile

```diff
@@ -82,13 +77,10 @@ test: prep
 		bundle exec --gemfile $$GEMFILE \
 		rspec spec
 
-push-image: prep
-	@docker push $(NAMESPACE)/r10k:$(VERSION)
-	@docker push $(NAMESPACE)/r10k:$(VERSION)-arm64
-ifeq ($(IS_LATEST),true)
-	@docker push $(NAMESPACE)/r10k:$(LATEST_VERSION)
-	@docker push $(NAMESPACE)/r10k:$(LATEST_VERSION)-arm64
-endif
+# this will use cached output from build target if it exists
+push-image: platform=linux/amd64,linux/arm64
+push-image: output_type=registry
+push-image: prep build
```

Comment:
I'm unclear on how pushing happens now..

Reply:
Yeah, so the `docker buildx build` cli is a bit obtuse. In the OG v1 Docker build process, you would:

* `docker build` - place the image build in local daemon (there's also a `--output` switch to further customize - see https://docs.docker.com/engine/reference/commandline/build/#custom-build-outputs). Side note - you can selectively enable / disable using the newer buildx builder through config.
* `docker push` - copy image from local daemon to registry somewhere

Our process to this point has been pretty similar, but we rely on use of `docker buildx` cli (confusingly the old `docker build` cli has a different set of options but can also run on top of the `buildx` engine)

* `docker buildx build` - use a buildx builder to place the image build in local daemon
* `docker push` - copy image from local daemon to docker hub

However, `docker buildx build` is really designed to be able to write the output of the build to even more [output locations](see https://docs.docker.com/engine/reference/commandline/buildx_build/#output):

* `local` (file system)
* `tar` (tarball)
* `oci` (tarball)
* `docker` (docker tarball, automatically loaded into the daemon)
* `image`
* `registry`

There are shorthand switches that you can use like `--load` (equivalent to `--output=type=docker`) and `--push` (equivalent to `--output=type=registry`). So the new process is:

* `docker buildx build --output=type=docker` to produce only the amd64 image for use in tests. It gets built by the builder and loaded into the local daemon.
* `docker buildx build --output=type=registry` to produce a multi-platform image and push it to docker hub. The previous amd64 layers are cached / reused (so that part should effectively be a no-op) in our pipelines, but the arm64 bits will be generated fresh.

This seems like the best blend of usability / sensible builds. I can add a few more comments if it makes it clearer.

One alternative I thought about was to always build both images, writing them to local disk, then loading in just the amd64 for tests. Then `build` is always used for building and `push-image` is always used for pushing cached artifacts... the only problem is that most environments will not have the tools installed to support multi-platform image building (i.e. it won't be easy for someone who pulls the repo to just run `make test`), so I opted for the approach that I did.

Reply:
That all makes sense to me, thanks so much for explaining!

Reply:
A comment noting that `--output=type=registry` is equivalent to `--push`, like you've done for `--load`, would probably be helpful!